app/main/views.py

```diff
@@ -47,8 +52,67 @@ def get_iaas():
     return resp
 
 
-@main.route('/services/<id>')
-def get_service(id):
-    service = Service.query.filter(Service.id == id).first_or_404()
+@main.route('/services', methods=['GET'])
+def list_services():
+    return jsonify(services=map(jsonify_service, Service.query.all()))
 
-    return jsonify(id=id, data=service.data)
+
+@main.route('/services', methods=['POST'])
+def add_service():
+    data = get_json_from_request()
+
+    service = Service(data=data['services'])
+    db.session.add(service)
+    db.session.commit()
+
+    return jsonify(services=jsonify_service(service)), 201
+
+
+@main.route('/services/<service_id>', methods=['PUT'])
+def update_service(service_id):
+    service = Service.query.filter(Service.id == service_id).first_or_404()
+    data = get_json_from_request()
+    if 'id' in data['services']:
+        if data['services'].pop('id') != service.id:
+            abort(400, "Invalid service ID provided")
+
+    service.data = data['services']
+
+    db.session.add(service)
+    db.session.commit()
+
+    return jsonify(services=jsonify_service(service)), 200
+
+
+@main.route('/services/<service_id>', methods=['GET'])
+def get_service(service_id):
+    service = Service.query.filter(Service.id == service_id).first_or_404()
+
+    return jsonify(services=jsonify_service(service))
+
+
+def jsonify_service(service):
+    data = dict(service.data.items())
+    data['id'] = service.id
+
+    data['links'] = [
+        {
+            "rel": "self",
+            "href": url_for('.get_service',
+                            service_id=service.id,
+                            _external=True)
+        }
+    ]
+    return data
+
+
+def get_json_from_request():
+    if request.content_type != 'application/json':
+        abort(400, "Unexpected Content-Type, expecting 'application/json'")
+    data = request.get_json()
+    if data is None:
+        abort(400, "Invalid JSON; must be a valid JSON object")
+    if 'services' not in data:
+        abort(400, "Invalid JSON must have a 'services' key")
```

Comment:
Are we requiring this for the POST as well as the responses? I will need to change the schemas if so, as they don't include the "services" wrapper.